homeassistant/components/openuv/coordinator.py

```diff
@@ -1,23 +1,94 @@
 """Define an update coordinator for OpenUV."""
 from __future__ import annotations
 
+import asyncio
 from collections.abc import Awaitable, Callable
 from typing import Any, cast
 
-from pyopenuv.errors import OpenUvError
+from pyopenuv.errors import InvalidApiKeyError, OpenUvError
 
-from homeassistant.core import HomeAssistant
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
 from homeassistant.helpers.debounce import Debouncer
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import LOGGER
+from .const import DOMAIN, LOGGER
 
 DEFAULT_DEBOUNCER_COOLDOWN_SECONDS = 15 * 60
 
 
+class InvalidApiKeyMonitor:
+    """Define a monitor for failed API calls (due to bad keys) across coordinators."""
+
+    DEFAULT_FAILED_API_CALL_THRESHOLD = 5
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self._count = 1
+        self._lock = asyncio.Lock()
+        self._reauth_flow_manager = ReauthFlowManager(hass, entry)
+        self.entry = entry
+
+    async def async_increment(self) -> None:
+        """Increment the counter."""
+        LOGGER.debug("Invalid API key response detected (number %s)", self._count)
+        async with self._lock:
+            self._count += 1
+            if self._count > self.DEFAULT_FAILED_API_CALL_THRESHOLD:
+                self._reauth_flow_manager.start_reauth()
+
+    async def async_reset(self) -> None:
+        """Reset the counter."""
+        async with self._lock:
+            self._count = 0
+            self._reauth_flow_manager.cancel_reauth()
+
+
+class ReauthFlowManager:
+    """Define an OpenUV reauth flow manager."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self.entry = entry
+        self.hass = hass
+
+    @callback
+    def _get_active_reauth_flow(self) -> FlowResult | None:
+        """Get an active reauth flow (if it exists)."""
+        try:
+            [reauth_flow] = [
+                flow
+                for flow in self.hass.config_entries.flow.async_progress_by_handler(
+                    DOMAIN
+                )
+                if flow["context"]["source"] == "reauth"
+                and flow["context"]["entry_id"] == self.entry.entry_id
+            ]
+        except ValueError:
+            return None
+
+        return reauth_flow
```

Comment:
This looks like we could move this into the base config_entries class as https://github.com/home-assistant/core/pull/81753 will make it do the same thing.

Reply:
Can you clarify the recommendation? Wait for #81753, roll forward with this, then do a follow-on PR with your second comment?

Reply:
Let's merge this now and and then in a followup it would be good to make a new function that in `homeassistant/config_entries.py` that takes 

```
        if any(
            flow
            for flow in hass.config_entries.flow.async_progress()
            if flow["context"].get("source") == SOURCE_REAUTH
            and flow["context"].get("entry_id") == self.entry_id
        )
```

and

```
        try:
            [reauth_flow] = [
                flow
                for flow in self.hass.config_entries.flow.async_progress_by_handler(
                    DOMAIN
                )
                if flow["context"]["source"] == "reauth"
                and flow["context"]["entry_id"] == self.entry.entry_id
            ]
        except ValueError:
            return None
```

and makes a new function in `homeassistant/config_entries.py` that yields matching flows (I probably should have left it as async_progress_by_handler as it would have been a bit more efficient) which both places can call instead of writing it out.